content/en/blog/_posts/2024-11-21-API-Streaming/2024-11-21-API-Streaming.md

```diff
@@ -0,0 +1,91 @@
+By streaming each item individually instead of returning the entire list, this method maintains constant memory overhead. 
+The server is bound by the maximum allowed size of an object in etcd plus a few additional allocations. 
+This approach drastically reduces the temporary memory usage compared to traditional LIST requests, ensuring a more efficient and stable system, 
+especially in clusters with a large number of objects of a given type or large average object sizes where despite paging memory consumption used to be high.
+
+Building on the insight gained from the synthetic test (see the synthetic test section for more details), we developed an automated performance test to systematically evaluate the impact of the WatchList feature. 
+This test replicates the same scenario, generating a large number of secrets with a large payload and scaling the number of informers to simulate heavy LIST request patterns. 
+The automated test is executed periodically to monitor memory usage of the server with the feature enabled and disabled.
+
+The results showed significant improvements with the WatchList feature enabled. 
+With the feature turned on, the kube-apiserver’s memory consumption stabilized at approximately **2 GB**. 
+In contrast, with the feature disabled, memory usage increased to approximately **20GB**, a **10x** increase! 
+These results confirm the effectiveness of the new streaming API, which reduces the temporary memory footprint.
+
+## Enabling API Streaming for your component
+
+Upgrade to Kubernetes 1.32. Make sure your cluster uses etcd in version 3.4.31+ or 3.5.13+. 
+Enable WatchListClient for client-go. For details on enabling the feature gate in client-go, please visit https://kubernetes.io/blog/2024/08/12/feature-gates-in-client-go. 
+
+## What's Next?
+In Kubernetes 1.32, the feature is enabled in kube-controller-manager by default despite its beta state. 
+This will be expanded to other core components like kube-scheduler or kubelet eventually, when it is promoted to GA the latest. 
+Other 3rd-party components are encouraged to opt-in to the feature during the beta phase, especially when they are at risk of accessing a large number of resources or kinds with potentially large object sizes.
+
+For the time being, [API Priority and Fairness](https://kubernetes.io/docs/concepts/cluster-administration/flow-control) assigns a reasonable small cost to LIST requests. 
+This is necessary to allow enough parallelism for the average case where LISTs are cheap enough. 
+But it does not match the spiky exceptional situation of many and large objects. 
+When WatchList is used by the majority of the ecosystem, the LIST cost estimation can be changed to larger values without risking degraded performance in the average case,
+and with that increasing the protection against this kind of requests that can still hit the apiserver in the future.
```

Comment:
Do you agree with:
```suggestion
API Priority and Fairness does not consider the size of collections when setting priorities or rate limits;
these controls happen before the size of the response is known.
The current small cost assigned for **list** requests allows enough parallelism for the average case
where **list** requests are cheap enough, nut it does not match the spiky exceptional situation of many
and / or large objects. 
Once the majority of the Kubernetes ecosystem has switched to watch lists, the **list** cost estimation can be changed to larger values without risking degraded performance in the average case,
and with that increasing the protection against this kind of requests that can still hit the API server in the future.
```
?

Reply:
>  API Priority and Fairness does not consider the size of collections when setting priorities or rate limits;

No - this isn't true. APF *does* consider size of collection, because we take number of object (per namespace or globally) of a given type when estimating cost of the request.

> The current small cost assigned for **list** requests allows enough parallelism for the average case

That's also not true - the cost pretty well reflects the CPU cost. It doesn't reflect cost of RAM.

Reply:
OK. Please fix the style issues, and if you want to explain the CPU vs. RAM detail, you're welcome to.

Reply:
**Revised suggestion**
```suggestion
API Priority and Fairness does not consider the size of objects in collections when setting priorities
or rate limits; these controls happen before the size of the response is known.
Once the majority of the Kubernetes ecosystem has switched to watch lists, the **list** cost estimation can be changed a larger multiplier without risking degraded performance in the average case,
and with that increasing the protection against this kind of requests that can still hit the API server in the future.
```
?